File: solutions/python3/0093.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        ret = []
        self.dfs(s, 0, 0, [""] * 4, ret)
        return ret

    def dfs(self, s: str, depth: int, index: int, curr: List[str], ret: List[str]) -> None:
        if depth == 4 and index == len(s):
            ret.append(".".join(curr))
            return
        if depth == 4 or index == len(s):
            return

        for i in range(1, 4):
            if index + i > len(s):
                return
            if i > 1 and s[index] == '0':
                return
            temp = s[index: index + i]
            if int(temp) > 255:
                return
            curr[depth] = temp
            self.dfs(s, depth + 1, index + i, curr, ret)
            curr[depth] = ""

File: solutions/cpp/0093.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> ans;
        vector<string> curr(4);
        dfs(s, 0, 0, curr, ans);
        return ans;
    }

    void dfs(string s, int depth, int index, vector<string>& curr, vector<string>& ans) {
        if (depth == 4 && index == s.size()) {
            ans.push_back(curr[0] + "." + curr[1] + "." + curr[2] + "." + curr[3]);
            return;
        }
        if (depth == 4 || index == s.size()) return;

        for (int i = 1; i <= 3; ++i) {
            if (index + i > s.size()) return;
            if (i > 1 && s[index] == '0') return;
            string temp = s.substr(index, i);
            if (stoi(temp) > 255) return;
            curr[depth] = temp;
            dfs(s, depth + 1, index + i, curr, ans);
            curr[depth] = "";
        }
    }
};